store/pegasus.go

```diff
@@ -0,0 +1,299 @@
+	PegasusTagPattern = "go_cache_tag_%s"
+	// Pegasus ttl(time-to-live) in seconds: -1 if ttl is not set; -2 if entry doesn't exist
+	PegasusNOTTL   = -1
+	PegasusNOENTRY = -2
+
+	DefaultTable             = "go_cache_pegasus"
+	DefaultTablePartitionNum = 4
+	DefaultScanNum           = 100
+)
+
+var (
+	// empty represent empty sort key, more info reference: https://github.com/XiaoMi/pegasus-go-client/blob/f3b6b08bc4c227982bb5b73106329435fda97a38/pegasus/table_connector.go#L83
+	empty = []byte("-")
+)
+
+// OptionsPegasus is options of Pegasus
+type OptionsPegasus struct {
+	Options
+	MetaServers []string
+
+	TableName         string
+	TablePartitionNum int
+	TableScanNum      int
+}
+
+// PegasusStore is a store for Pegasus
+type PegasusStore struct {
+	client  pegasus.Client
+	options *OptionsPegasus
+}
+
+// NewPegasus creates a new store to pegasus instance(s)
+func NewPegasus(options *OptionsPegasus) (*PegasusStore, error) {
+	if options == nil {
+		options = &OptionsPegasus{}
+	}
+
+	if err := createTable(options); err != nil {
+		return nil, err
+	}
+
+	client := pegasus.NewClient(pegasus.Config{
+		MetaServers: options.MetaServers,
+	})
+	table, err := client.OpenTable(context.Background(), options.TableName)
+	defer table.Close()
+	if err != nil {
+		return nil, err
+	}
+
+	return &PegasusStore{
+		client:  client,
+		options: options,
+	}, nil
+}
+
+// validateOptions validate pegasus options
+func validateOptions(options *OptionsPegasus) error {
+	if len(options.MetaServers) == 0 {
+		return errors.New("pegasus meta servers must fill")
+	}
+	if len(options.TableName) == 0 {
+		options.TableName = DefaultTable
+	}
+	if options.TablePartitionNum < 1 {
+		options.TablePartitionNum = DefaultTablePartitionNum
+	}
+	if options.TableScanNum < 1 {
+		options.TableScanNum = DefaultScanNum
+	}
+
+	return nil
+}
+
+// createTable for create table by options
+func createTable(options *OptionsPegasus) error {
+	if err := validateOptions(options); err != nil {
+		return err
+	}
+
+	tableClient := admin.NewClient(admin.Config{MetaServers: options.MetaServers})
+	tableList, err := tableClient.ListTables(context.Background())
+	if err != nil {
+		return err
+	}
+
+	for i := range tableList {
+		if tableList[i].Name == options.TableName {
+			return nil
+		}
+	}
+
+	// if not found then create table of options
+	return tableClient.CreateTable(context.Background(), options.TableName, options.TablePartitionNum)
+}
+
+// dropTable for drop table
+func dropTable(options *OptionsPegasus) error {
+	tableClient := admin.NewClient(admin.Config{MetaServers: options.MetaServers})
+	return tableClient.DropTable(context.Background(), options.TableName)
+}
+
+// Close when exit store
+func (p *PegasusStore) Close() error {
+	return p.client.Close()
+}
+
+// Get returns data stored from a given key
+func (p *PegasusStore) Get(key interface{}) (interface{}, error) {
+	table, err := p.client.OpenTable(context.Background(), p.options.TableName)
```

Comment:
Note to myself: I will pass parent context to these methods directly

Reply:
done!